### ISSUE TITLE
"too many open files" when pushing a large number of modified files

### ISSUE THREAD

**Issue by pez252** [odeke-em/drive#172]
TLDR; Files opened by drive for uploading because they were modified are not closed after upload.

I have just started using drive after discovering it yesterday and wasn't able to find anything about ran into an issue when pushing a large number of files.

The push scans for new/removed/modified files and prompts. After confirming that I would like the modifications pushed it begins uploading the files. After running for a while I begin receiving errors like:

/path/to/the/file/file.name: open /home/username/gdrive/path/to/the/file/file.name: too many open files

On my initial run I had 4738 files that were modified. It ran to 16.92% before complaining about open files.
Running drive again against the same path there were 3721 modified files remaining. This run went to 27.47%.
On the third run there were 2704 modified files remaining. This run went to 16.6%

I did some checking with lsof during the third run. During the scanning phase I saw at most 9 files open by the drive process (only counting the files that were being sync'd). When drive was waiting at the prompt to confirm the push changes no files to sync were open, only dependencies of drive. As soon as the sync began files being sync'd started being listed but were never closed. lsof -p [drive's PID] | wc -l grew quickly to 80 then 200 then 950 open files as files.

I tested doing a push with a large number of new files (15000) and did not run into this issue. It seems to be isolated to pushing modified files.

**Comment by pez252** [comment]
Additional info...

result of cat /proc/version: `Linux version 3.11.0-12-generic (buildd@allspice) (gcc version 4.8.1 (Ubuntu/Linaro 4.8.1-10ubuntu7) ) #19-Ubuntu SMP Wed Oct 9 16:20:46 UTC 2013`
Installed with `go get -u github.com/odeke-em/drive/cmd/drive`
result of go version: `go version go1.1.2 linux/amd64`

go was installed from a default repo. To install newer go I did:
```
go get gopkg.in/niemeyer/godeb.v1/cmd/godeb
sudo apt-get remove golang
sudo apt-get remove golang-go
godeb install 1.3.3
```
new result of go version: `go version go1.3.3 linux/amd64`

Re-ran my tests from above and still have the issue with drive push a/path/to/stuff not closing files when the files being uploaded were selected because they were modified.

**Comment by pez252** [comment]
After retrying a push with a large set of files (~15000) that did not exist remotely I received a fatal error from go. This lead me to believe there was an issue with my go environment after installing the newer version above. I did the following to cleanup:
```
godeb remove
sudo apt-get purge golang*
rm -rf  ~/.go
godeb install 1.2.2
go get github.com/rakyll/drive/cmd/drive
mv ~/gdrive ~/gdrive_old
drive init ~/gdrive
ln -s /actual/path/to/data/ ~/gdrive/path
```

After this it seems my issue is gone... Running a push with a large number of modified files works. Running a push with a large number of new files works.

Lesson learned... don't assume the version in the default repositories is remotely current.

**Comment by pez252** [comment]
Re-opened issue for a follow up that may be related to some of the install/re-install things done above. Now that drive push seems to be working, I no longer receive a summary of files to be updated at the confirmation prompt. Related to the above?

If not, feel free to re-close this.

**Comment by odeke-em** [comment]
Hello @pez252 sorry for the late reply, been swamped the whole morning and afternoon. Welcome to drive!
As per your issue disappearing:
```
go get github.com/rakyll/drive/cmd/drive
```
wrong installation please make sure you are doing installs from this repo
```shell
$ go get github.com/odeke-em/drive/cmd/drive
```

Also the last time I checked the number of open files is set by kernel and if you have way too many applications open e.g web browser tabs, terminals, pushes etc in this case you you might have tens of thousands of files open but your system has only few resources this trips things out. Please Google for `ulimit` and also how to find the max number of open files.
I can't reply much right now but when I get home I'll try to.
Thank you for reporting this.

**Comment by pez252** [comment]
I did indeed put the wrong github repo when re-installing. I ran some additional tests worried that the issue with files not being closed would come back after re-installing the correct version of drive. The issue has not recurred, and the change summary is now displayed.  It seems that the problem was with drive and 1.1.2 version of go. Since 1.2+ is listed as a dependency I am closing this issue. Thanks for the assistance, and thanks for the work on this great tool.

**Comment by odeke-em** [comment]
Thank you too for the analysis, using drive and for raising this issue. It is indeed something to watch out for when doing file uploads. Please don't hesitate to reach out in case of any issues or ideas.
